Implement lock on suspend

apaan opened the pull request:
This patch fixes two issues #145 and #175. 

It is adopted from xfce4-screensaver patch:
https://git.xfce.org/apps/xfce4-screensaver/commit/?id=9e53bb2866

Introduced new Inhibitor lock to make sure lock was in place and
user will be greeted with unlock dialog when resumes from suspend/
hibernate.

lukefromdc commented:
Can users turn this on and off with a gsetting? Not everyone wants the screen locked when coming out of suspend. Some need is and some do not.

The obvious example of this wouldbe many users of encrypted laptops, who must maintain absolute custody of the machine anyway whenever it is booted as cold boot attacks and other means can recover keys from encrypted computers stolen/confiscated while running and decrypted.

apaan commented:
Testing on my laptop, without this patch, mate-screensaver behaves according to option "Lock screen when screensaver is active" in Screensaver Preferences (stored in ```lock-enabled``` gsetting):
- when the option is active, the screen will lock on suspend/hibernate (screensaver is activated too)
- when the option is inactive, the screen will not lock on suspend (screensaver is not activated, even if the suspend time is longer than "Idle time" option)

If this is correct, i will modify the patch so that the behavior will not change.

Thanks

lukefromdc commented:
The described behavior in master "Lock screen when screensaver is active" should be the correct behavior and is simpler than adding a new and separate action.

monsta commented on src/gs-listener-dbus.c:
```diff
@@ -1656,6 +1751,27 @@ listener_dbus_handle_system_message (DBusConnection *connection,
+				if (active) {
+					gs_debug ("Logind wanted to sleep");
+					g_signal_emit (listener, signals[LOCK], 0);
+					g_signal_emit (listener, signals[SHOW_MESSAGE], 0, NULL, NULL, NULL);
```
What does `SHOW_MESSAGE` signal do in this case...?

apaan replied:
To bring up unlock dialog, because we don't have "show unlock" signal.
This is done twice (one before suspend, and one after suspend) because there is a probability that the suspend process is taking long time, and the dialog is closed before suspend. (see issue #175 )

monsta commented:
Xfce's screensaver is a fork of this one... how does it work without that new inhibitor?

monsta commented:
> The described behavior in master "Lock screen when screensaver is active" should be the correct behavior and is simpler than adding a new and separate action.

I recall we have a bunch of "lock-on-something" keys in mate-power-manager schema. However, it's probably only related to cases when suspend is activated on (hardware) button press or on closing the laptop lid...

apaan commented:
> Xfce's screensaver is a fork of this one... how does it work without that new inhibitor?

Actually I started by copying the Xfce's patch (I even put a link to it in the commit message). But it didn't always work on my notebook. So i dug logind documentation and found this (taken from https://www.freedesktop.org/wiki/Software/systemd/inhibit/):

> Note that watching PrepareForShutdown(true)?/PrepareForSleep(true) without taking a delay lock is racy and should not be done, as any code that an application might want to execute on this signal might not actually finish before the suspend/shutdown cycle is executed. Again: if you watch PrepareForSuspend(true), then you really should have taken a delay lock first.

apaan commented:
> > The described behavior in master "Lock screen when screensaver is active" should be the correct behavior and is simpler than adding a new and separate action.
> 
> I recall we have a bunch of "lock-on-something" keys in mate-power-manager schema. However, it's probably only related to cases when suspend is activated on (hardware) button press or on closing the laptop lid...

Just checked that. mate-power-manager has ```lock-use-screensaver```: Whether to use the screen lock setting of mate-screensaver to decide if the screen is locked after a hibernate, suspend or blank screen.

Current behavior is different from master in this regard. Will look into it later

monsta commented:
Ok, so we delay the suspend until we display the lock dialog, and then release it...

> Note that watching PrepareForShutdown(true)?/PrepareForSleep(true) without taking a delay lock is racy and should not be done, as any code that an application might want to execute on this signal might not actually finish before the suspend/shutdown cycle is executed. Again: if you watch PrepareForSuspend(true), then you really should have taken a delay lock first.

So that's why we also acquire the lock after resuming? To play it safe for the next PrepareForSleep...

monsta commented:
Interesting that Xfce's screensaver still doesn't have neither inhibit nor showing the dialog before the suspend...

raveit65 commented:
> So that's why we also acquire the lock after resuming? To play it safe for the next PrepareForSleep...

Why not simply testing the PR, it's self explanatory....

apaan commented:
> > The described behavior in master "Lock screen when screensaver is active" should be the correct behavior and is simpler than adding a new and separate action.
> 
> I recall we have a bunch of "lock-on-something" keys in mate-power-manager schema. However, it's probably only related to cases when suspend is activated on (hardware) button press or on closing the laptop lid...

Pushed a change: This patch now lets mate-power-manager to lock or not. This patch only ensures that the first thing that we see on resume is unlock dialog. Same behavior as before, but now, this patch behaves correctly on hidden mate-power-manager settings.

monsta commented on src/gs-listener-dbus.c:
```diff
@@ -1619,6 +1621,66 @@ properties_changed_match (DBusMessage *message,
+	dbus_error_init (&error);
+
+	message = dbus_message_new_method_call (SYSTEMD_LOGIND_SERVICE,
+											SYSTEMD_LOGIND_PATH,
```
Minor thing: indentation is way off...

apaan replied:
ouch. Thanks. Failed to see that the file is indented by tabs. Will correct it in the next patch (with spaces)

monsta commented on src/gs-listener-dbus.c:
```diff
@@ -1656,6 +1718,27 @@ listener_dbus_handle_system_message (DBusConnection *connection,
+				g_signal_emit (listener, signals [PREPARE_FOR_SLEEP], 0, active);
+				if (listener->priv->logind_inhibit_lock) {
+					gs_debug ("Releasing inihibitor lock");
+					close (listener->priv->logind_inhibit_lock);
```
Hmm... `close` won't set the descriptor to zero.

apaan replied:
Thanks

monsta commented on src/gs-monitor.c:
```diff
@@ -280,6 +280,17 @@ static void listener_simulate_user_activity_cb(GSListener* listener, GSMonitor*
 
+static void listener_prepare_for_sleep_cb(GSListener* listener, gboolean prepare, GSMonitor* monitor)
+{
+	gboolean locked;
```
Should we also check if `prepare` is `TRUE` here?

apaan replied:
As with previous discussion int he PR, it is possible that the unlock dialog has already been closed again before suspend, it the system was slow. This patch deliberately send show message after suspend as well, so that it will be shown again.

The callback signature mimics "PrepareForSleep" logind message, so it has prepare parameter (renamed from ```active``` for clarty). Maybe some time in the future we will need that for another use.

monsta replied:
Ok, makes sense.

raveit65 commented:
@apaan 
I found the working version of your work.
https://github.com/mate-desktop/mate-screensaver/compare/410fcb6fcbff537ba0cb24b834a86c68721cdb7d..4dbbbe7b8e72e3a54b5d0657a94b794bcee1ab04.patch
Thanks for that :)
I will use that for fedora and step back from the review here, because I have no idea why reviewers breaks your work here with their ideas....

vkareh commented on src/gs-listener-dbus.c:
```diff
@@ -1619,6 +1621,66 @@ properties_changed_match (DBusMessage *message,
+						DBUS_TYPE_STRING, &who,
+						DBUS_TYPE_STRING, &why,
+						DBUS_TYPE_STRING, &mode,
+						DBUS_TYPE_INVALID) == FALSE)
```
Indentation here is off as well

apaan replied:
thanks

vkareh commented on src/gs-listener-dbus.c:
```diff
@@ -1619,6 +1621,66 @@ properties_changed_match (DBusMessage *message,
+	}
+
+	reply = dbus_connection_send_with_reply_and_block (connection, message,
+		-1, &error);
```
Indentation...

apaan replied:
Thanks

vkareh commented:
Still need to figure out why it doesn't work for @raveit65 - that's a potential blocker...

monsta commented:
This also works for me if I disable locking in screensaver settings, but explicitly tell the power manager to ask the screensaver to lock on suspend.

```
$ gsettings get org.mate.screensaver lock-enabled
false
$ gsettings get org.mate.power-manager lock-use-screensaver
false
$ gsettings get org.mate.power-manager lock-suspend
true
```
Suspend was tested by setting the power button to perform suspend in power manager settings, then pressing the power button (obviously).

monsta commented:
The commit message should probably be changed, because lock-on-suspend is already implemented, and this change just ensures it...

raveit65 commented:
Ok, i was able to reproduce the issue and i don't think it is related to PR.
On a system with 2 monitors (non laptop) is is possible that the lock-dialog is on a monitor which isn't powered on after coming from suspend-to-ram = ***Computer are stupid!***
Than you see only a black screen at the monitor which is on.
In details:
1. Both monitors are enabled in mate-window-properties
2. Suspend the system
3. Plug off the second monitor from electrical net to save power.
4. Go shopping  (forget everything!)
5. Come back from shopping, and start system from suspend with one monitor.
6. Have a black screen without a lock-dialog.

Note, those steps aren't always reproducible and i think this is a monitor detection issue with a multi-monitor setup, not related to mate-screensaver.

apaan commented:
Force-pushed a new version of this patch:
- Fixed indentation problems (thanks @vkareh, @monsta)
- Fixed not setting ```logind_inhibit_lock``` to zero after closing it (thanks @monsta)
- Update commit message
